File: 30 Dec Add Number Linked Lists/code.cpp
/*
class Node {
  public:
    int data;
    Node* next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};
*/

class Solution {
  public:
    Node* rec(Node*h){
        Node*p=NULL,*nx=NULL;
        
        while(h){
            nx=h->next;
            h->next=p;
            p=h;
            h=nx;
        }
        return p;
    }
    Node* addTwoLists(Node* head1, Node* head2) {
        // code here
        Node*res=new Node(-1);
        head1=rec(head1),head2=rec(head2);
        int c=0;
        Node* t=res;
        while(head1 || head2 || c){
            int v=((head1?head1->data:0) + (head2?head2->data:0) + c);
            c=v/10;
            t->next=new Node(v%10);
            t=t->next;
            if(head1)head1=head1->next;
            if(head2)head2=head2->next;
        }
        res=rec(res->next);
        while(res->next and res->data==0)res=res->next;
        return res;
    }
};

File: 31 Dec Palindrome Linked List/code.py


class Solution:
    def rec(self,h):
        if(not h or not h.next):return h
        pr,nx=None,None
        while(h):
            nx=h.next
            h.next=pr
            pr=h
            h=nx
        return pr
    def isPalindrome(self, head):
        if(not head.next):
            return True
        sl,fs=head,head
        while(fs and fs.next):
            sl=sl.next
            fs=fs.next.next
        sl=self.rec(sl)
        fs=head
        while(fs and sl):
            if(fs.data!=sl.data):return False
            fs,sl=fs.next,sl.next
        return True
        
        